### DISCUSSION TITLE
Manage Saral ML Assets better with Firebase ML and build options

### DISCUSSION THREAD

**Discussion by dileep-gadiraju** [Sunbird-Saral/Community#66] [General]
Hi @pramodkvarma / @surendrasinghs,

cc: @arman-tdev , @apurvabisht97 , @venky3692 

To manage Saral ML Assets better with Firebase ML and build options, we are working on the below feature. Please review and let us know if you have any input on this.

**Highlights:**
1) Implementor will have the option to choose the ML Assets required while packaging. 
   This will also save App size by excluding unwanted ML Assets for a given implementation.
2) Decouple App from ML Assets. This enables the implementor to keep updating the re-trained models from time to time without re-publishing the app into Play Store.

**Considerations**
1) When to download the model from firebase to mobile? Only when the mobile is on wifi network?
2) Make sure this feature works on all mobiles where the current version app is supported.
3) Make sure this feature can be enabled/disabled with a simple configuration
4) If a model is already present in the app, override the latest model from the firebase ML service.
5) Make sure the model downloaded from Firebase ML is one-time and the app can operate in offline mode thereafter.
6) If a model that supports a specific extraction method is not present in Firebase ML and in the installed app, show an alert message when the user tries to scan the layout.

Reference: https://firebase.google.com/docs/ml/android/use-custom-models

We are doing a POC to address the above challenges/considerations. Once these aspects are proven, we will introduce this as a feature.
https://project-sunbird.atlassian.net/browse/SS-42 (POC)
https://project-sunbird.atlassian.net/browse/SS-95 (Full Feature)

**Comment by surendrasinghs** [comment]
Thanks for bringing this up. App size optimization is very much needed, considering the target audience & devices where Saral reference app gets installed and used.

Please see if the following helps

1. When the user downloads the app for the first time, they would need to download the model - whether on Wi-Fi or mobile data, right? 
2. I believe, without this the app cannot proceed/function, right?
3. The choice we can provide to users - whether they want to download or cancel. Cancelling will exit the app, assuming # 2 is true.
4. On any subsequent updates/changes to the model, users should be prompted to download - they can skip and download later, yet continue to use the app - unless the functionality/scan cannot be completed without a model upgrade/update.
5. Any subsequent app/container update should not unnecessarily result in model update/download. Models already downloaded should be available in newer versions of the app.

Hope this helps. Let's discuss this over a quick call if need be.

**Reply by dileep-gadiraju** [reply]
thanks for the comments @surendrasinghs.  We are also giving options for the implementor to include or exclude a model based on the implementation use cases. Today we have 2 ML Assets (Numeric, Alpha-numeric), so the implementor can choose to include only the Numeric model and exclude Alpha-numer for u.p implementation for example. 

Also, the user prompt to download the model we will consider as part of phase 2 implementation once minimal features are delivered and stabilized. 


@arman-tdev, also include wifi check as firebase remote config flag instead of hard wiring it.
